docs/extensibility/custom-resources.md

```diff
@@ -0,0 +1,432 @@
+
+```csharp
+var builder = DistributedApplication.CreateBuilder(args);
+var maildev = builder.AddMailDev("maildev");
+builder.AddProject<Projects.MailDevResource_NewsletterService>("newsletterservice")
+       .WithReference(maildev);
+
+builder.Build().Run();
+```
+
+After updating the `Program.cs` source file launch the AppHost again and verify that the .NET project that was added started and that the environment variable `ConnectionStrings__maildev` was added to the process (click on the _Details_ link on the `newsletterservice` row in the dashboard).
+
+![Environment variables for Newsletter Service in .NET Aspire Dashboard](./media/maildev-envvar.png)
+
+### Step 4b: Use connection string to send message.
+
+To use the SMTP connection details that were injected into the newsletter service project we will inject an instance of <xref:System.Net.Mail.SmtpClient> into the dependency injection container as a singleton. Add the following code to the `Program.cs` file in the `MailDevResource.NewsletterService` project to setup the singleton service.
+
+```csharp
+builder.Services.AddSingleton<SmtpClient>((sp) =>
+{
+    var smtpUri = new Uri(builder.Configuration.GetConnectionString("maildev")!);
+    var smtpClient = new SmtpClient(smtpUri.Host, smtpUri.Port);
+    return smtpClient;
+});
+```
+
+To test the client we will add two simple subscribe and unsubscribe GET methods to the newsletter service. Add the following code to the `Program.cs` file in the `MailDevResource.NewsletterService` project to setup the ASP.NET Core routes:
+
+```csharp
+app.MapGet("/subscribe", async (SmtpClient smtpClient, string email) => {
+    var message = new MailMessage("newsletter@yourcompany.com", email);
+    message.Subject = "Welcome to our newsletter!";
+    message.Body = "Thank you for subscribing to our newsletter!";
+    await smtpClient.SendMailAsync(message);
+});
+
+app.MapGet("/unsubscribe", async (SmtpClient smtpClient, string email) => {
+    var message = new MailMessage("newsletter@yourcompany.com", email);
+    message.Subject = "You are unsubscribed from our newsletter!";
+    message.Body = "Sorry to see you go. We hope you will come back soon!";
+    await smtpClient.SendMailAsync(message);
+});
+```
+
+Once the `Program.cs` file is updated launch the AppHost and use your browser, or `curl` to hit the following URLs:
+
+```bash
+# Remember to replace the port to match your local environment.
+curl https://localhost:7289/subscribe?email=test@test.com
+curl https://localhost:7289/unsubscribe?email=test@test.com
+```
+
+If those API calls return a successful response then you should be able to click on the `maildev` resource the dashboard and the MailDev UI will show the emails that have been sent to the SMTP endpoint.
+
+![E-mails visible in the MailDev UI](./media/maildev-emails.png)
+
+## Technical details
+
+### IValueProvider and IManifestExpressionProvider
```

Comment:
```suggestion
### ReferenceExpression
```

I would focus this part of `ReferenceExpression` since you use that type directly. Then we can write a more detailed piece on the underlying interfaces.

Reply:
Is this more about the title of the section or did you want content changes. I do mention the interfaces but don't go into a whole heck of a lot of detail.